llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -4140,6 +4266,165 @@ bool AIE2InstructionSelector::select512BitG_AIE_STORE_SRS(
+  return MI;
+}
+
+// If two 256-bit SRS are input to 512-bit STORE through CONCAT, these are
+// combined into VST.SRS
+bool AIE2InstructionSelector::selectG_AIE_STORE_CONCAT_SRS(
+    MachineInstr &StoreI, MachineRegisterInfo &MRI) {
+
+  Register ConcatResult = (StoreI.uses().begin())->getReg();
+  MachineInstr *ConcatOp = MRI.getUniqueVRegDef(ConcatResult);
+
+  assert(ConcatOp && "Expected SSA.");
+
+  if (getLoadStoreSize(StoreI) != 512 ||
+      ConcatOp->getOpcode() != AIE2::G_INTRINSIC ||
+      cast<GIntrinsic>(*ConcatOp).getIntrinsicID() !=
+          Intrinsic::aie2_concat_I512_I256)
+    return false;
+
+  MachineInstr *SrsOpLo = stripBitCast(ConcatOp->getOperand(2).getReg(), MRI);
+  MachineInstr *SrsOpHi = stripBitCast(ConcatOp->getOperand(3).getReg(), MRI);
+
+  if (!canCombineSRSUPS(StoreI, *SrsOpLo) ||
+      !canCombineSRSUPS(StoreI, *SrsOpHi))
+    return false;
+
+  std::optional<AddressingModeInfo> AMI =
+      getOrDefineAddressingRegister(StoreI, MRI);
+  if (!AMI)
+    return false;
+
+  Register SrcRegLo = SrsOpLo->getOperand(2).getReg();
+  Register ShftRegLo = SrsOpLo->getOperand(3).getReg();
+  Register SignRegLo = SrsOpLo->getOperand(4).getReg();
+  Register SrcRegHi = SrsOpHi->getOperand(2).getReg();
+  Register ShftRegHi = SrsOpHi->getOperand(3).getReg();
+  Register SignRegHi = SrsOpHi->getOperand(4).getReg();
+
+  std::optional<LoadStoreOpcodes> LSOLo;
+  bool ConstantSignLo = false;
+  if (auto SignVal = getIConstantVRegValWithLookThrough(SignRegLo, MRI)) {
+    // SignVal = 1 for signed and 0 for dynamically signed
+    LSOLo = getCombinedOpcodeSRSUPS(StoreI, *SrsOpLo, AMI->ImmediateOffset,
+                                    SignVal.value().Value == 0x1);
+    ConstantSignLo = true;
+  } else {
+    LSOLo =
+        getCombinedOpcodeSRSUPS(StoreI, *SrsOpLo, AMI->ImmediateOffset, false);
+  }
+
+  assert(LSOLo && "Unexpected VST.SRS combine failure");
+
+  std::optional<LoadStoreOpcodes> LSOHi;
+  bool ConstantSignHi = false;
+  if (auto SignVal = getIConstantVRegValWithLookThrough(SignRegHi, MRI)) {
+    // SignVal = 1 for signed and 0 for dynamically signed
+    LSOHi = getCombinedOpcodeSRSUPS(StoreI, *SrsOpHi, AMI->ImmediateOffset,
+                                    SignVal.value().Value == 0x1);
+    ConstantSignHi = true;
+  } else {
+    LSOHi =
+        getCombinedOpcodeSRSUPS(StoreI, *SrsOpHi, AMI->ImmediateOffset, false);
+  }
+
+  assert(LSOHi && "Unexpected VST.SRS combine failure");
+
+  switch (AMI->MemI.getOpcode()) {
+  case AIE2::G_STORE:
+  case AIE2::G_AIE_POSTINC_STORE:
+  case AIE2::G_AIE_POSTINC_2D_STORE:
+  case AIE2::G_AIE_POSTINC_3D_STORE: {
+    auto StoreHigher = MIB.buildInstr(*LSOHi->OffsetOpcode)
+                           .addReg(AMI->PtrOp.getReg())
+                           .addImm(32) // Offset
+                           .addReg(SrcRegHi)
+                           .addReg(ShftRegHi);
+
+    auto StoreLower = MIB.buildInstr(LSOLo->ISelOpcode);
```

Comment:
I guess we are building the `vst.srs` instructions at the position of the `G_STORE` instruction. As a consequence, this means we will read the `crSat` and `crRnd` registers at that point, instead of where the `vsrs` instructions used to be. If those registers are changed in between (e.g. with `set_sat()`), this means `vst.srs` would be configured differently compared to the initial `vsrs`. Do you know if there is anything protecting us against that? Maybe we check for sideEffect when trying to move vsrs down?